pyrobosim/pyrobosim/navigation/execution.py

```diff
@@ -205,6 +251,65 @@ def validate_remaining_path(self) -> None:
 
             time.sleep(max(0, self.validation_dt - (time.time() - start_time)))
 
+    def detect_closed_hallway(self) -> None:
+        """
+        Get lidar measurement and determine if it's scanning a hallway.
+        If yes, it would update the robot recorded_closed_hallways knowledge.
+        It either remove (if it detects hallway is open) or add (if it detects hallway is close) the hallway
+        into the robot's knowledge.
+        """
+        if (self.robot is None) or (self.robot.world is None):
+            return
+
+        lidar_sensor = self.robot.sensors.get(self.lidar_sensor_name)  # type: ignore[arg-type, union-attr]
+        measured_angles = lidar_sensor.angles  # type: ignore[union-attr]
+
+        while self.following_path and (not self.abort_execution):
+            start_time = time.time()
+            cur_pose = self.robot.get_pose()
+
+            measured_lengths = lidar_sensor.get_measurement()  # type: ignore[union-attr]
+            analyse_pose = []
+            for angle, length in zip(measured_angles, measured_lengths):
+                if length < lidar_sensor.max_range_m:  # type: ignore[union-attr]
+                    # There are objects in lidar line of sight
+                    adjusted_angle = angle + cur_pose.get_yaw()
+                    x = cur_pose.x + length * np.cos(adjusted_angle)
+                    y = cur_pose.y + length * np.sin(adjusted_angle)
+                    analyse_pose.append((x, y))
+
+            for pose in analyse_pose:
+                for hallway in self.robot.world.hallways:
+                    # Check if there are pose intersecting with hallway polygon
+                    if shapely.intersects_xy(
+                        hallway.internal_collision_polygon, pose[0], pose[1]
+                    ):
```

Comment:
wait a second... this can be sped up dramatically too instead of a for-loop!

You can instead ignore the creation of `analyse_poses` and do something like this:
https://github.com/sea-bass/pyrobosim/blob/60a2da318dca039d7bd7bdc09f15d9bf59539259/pyrobosim/pyrobosim/sensors/lidar.py#L85-L91

Also I suspect this actually has a serious bug. `pose` only contains the endpoint of the lidar ray, meaning that if the lidar goes completely through a hallway and is on the other side, the line segment is ignored and it shows as collision free.

So yes, consider doing the intersection with all the lidar line segments.

Reply:
Correction here: you don't want to use transformed `self.orig_lidar_lines` here, but actually `self.lidar_lines`.

This will already give you all the real lidar sensor rays, so you just check that multi-line polygon's intersection with each hallway and you should be good to go.

As a bonus speedup, the list of hallways to check could be reduced. Technically if you keep track of all the hallways where the robot's knowledge *disagrees* with the ground truth, you should only have to do this operation with that subset of hallways instead of looping through all of them.